src/modules/recorder.module.ts

```diff
@@ -0,0 +1,373 @@
+    // If user is speaking already, return (this event can get fired twice)
+    if (userState.get(user.id) && speaking.has(Speaking.FLAGS.SPEAKING)) {
+        return
+    }
+
+    const channelData = voiceChannels.get(channelId)
+    if (!channelData) {
+        return
+    }
+
+    const connection = channelData.connection
+    const buffer = channelData.buffer
+
+    // If a user starts speaking, 
+    if (speaking.has(Speaking.FLAGS.SPEAKING)) {
+        userState.set(user.id, true)
+        const audio = connection.receiver.createStream(user, {mode: "pcm"})
+        const writer = createPCMBufferWriter(buffer)
+        audio.on("data", (chunk: Buffer) => {
+            addToBuffer(writer, chunk)
+        })
+    }
+
+    // If a user stops speaking, reset the flag
+    else if (userState.get(user.id) && !speaking.has(Speaking.FLAGS.SPEAKING)) {
+        userState.set(user.id, false)
+    }
+}
+
+// Join a voice channel and register listeners on user voice connections.
+async function joinChannel(channel: VoiceChannel) {
+    if (voiceChannels.has(channel.id)) {
+        return
+    }
+
+    const connection: VoiceConnection = await channel.join()
+    const buffer: PCMBuffer = createPCMBuffer(BUFFER_LEN_SEC)
+
+    // Play joining audio so Discord starts sending us data
+    connection.play("./yo.wav")
+    
+    // Register listener
+    connection.on("speaking", (user, speaking) => {
+        onUserChangeState(channel.id, user, speaking)
+    })
+
+    let lastCleared: number = timeToOffset(buffer, CLEAR_INTERVAL_SEC)
+    // Start interval to clear buffer
+    const interval: NodeJS.Timeout = setInterval(() => {
+        const now: number = timeToOffset(buffer, CLEAR_INTERVAL_SEC)
+        clearBuffer(buffer, lastCleared, now)
+        lastCleared = now
+    }, CLEAR_INTERVAL_SEC * MILLIS)
+
+    voiceChannels.set(channel.id, {
+        connection: connection,
+        buffer: buffer,
+        interval: interval,
+        lastClipTime: new Date(0)
+    })
+}
+
+// Leave a voice channel and clean up connections / intervals
+async function leaveChannel(channel: VoiceChannel) {
+    // always fetch the channel via API as using data from the cache
+    // may lead to the bot being stuck in an empty channel
+    const fetchedChannel: VoiceChannel = await channel.fetch(true) as VoiceChannel
+    const voiceChannelData = voiceChannels.get(channel.id)
+    if (voiceChannelData && fetchedChannel.members.size == 1) {
+        const connection: VoiceConnection = voiceChannelData.connection
+        voiceChannelData.connection.disconnect()
+        clearInterval(voiceChannelData.interval)
+        voiceChannels.delete(channel.id)
+    }
+}
+
+function randomClipId() {
+    let id: string = ""
+    for (let i = 0; i < CLIP_ID_LEN; i++) {
+        id += CLIP_ID_CHARS[Math.floor(Math.random() * CLIP_ID_CHARS.length)]
+    }
+    return id
+}
+
+async function saveClip(channel: VoiceChannel, clipName: string, textChannel: TextChannel) {

```

Comment:
Because you never use `await` in the scope of this function (not in subscopes of other functions/callbacks inside this functions), this does not need to be `async`. If that changes based on other comments, you can leave this as `async`.